homeassistant/components/vacuum/xiaomi_miio.py

```diff
@@ -71,6 +75,13 @@
         vol.All(vol.Coerce(int), vol.Clamp(min=-179, max=179)),
     vol.Optional(ATTR_RC_DURATION): cv.positive_int,
 })
+SERVICE_SCHEMA_CLEAN_ZONE = VACUUM_SERVICE_SCHEMA.extend({
+    vol.Required(ATTR_ZONE_ARRAY):
+        vol.All(list, [vol.ExactSequence([int, int, int, int])]),
+    vol.Exclusive(ATTR_ZONE_REPEATER, 'TimeToRepeats'):
+        vol.All(vol.Coerce(int), vol.Clamp(min=1, max=3)),
+    vol.Exclusive(ATTR_ZONE_REPEATER_TEMPLATE, 'TimeToRepeats'): cv.template,
```

Comment:
Why can't we use a service data_template?

Reply:
Would you explain to me how?

Reply:
maybe in next release. thanks

Reply:
We won't merge this with the current service schema if it's not required.

Reply:
Please help me to configure data_template.
I don't know how to configure it in my py code.
Thanks

Reply:
What do you need help with? Writing an automation using a data_template to test this PR? Did you test what I posted above?

Reply:
if I use data_template I get the following error: _Error while executing automation automation.test_vacuum_zone3. Invalid data for call_service at pos 1: expected int @ data['zone'][0]_

Reply:
Try adding a template with  `int` filter for the zones too.

Reply:
my new validator schema is:
```
ATTR_ZONE_ARRAY = 'zone'
ATTR_ZONE_ARRAY_TEMPLATE = 'zone_template'
ATTR_ZONE_REPEATER = 'repeats'
ATTR_ZONE_REPEATER_TEMPLATE = 'repeats_template'

SERVICE_SCHEMA_CLEAN_ZONE = VACUUM_SERVICE_SCHEMA.extend({
    vol.Exclusive(ATTR_ZONE_ARRAY, 'ZoneToClean'):
        vol.All(list, [vol.ExactSequence([int, int, int, int])]),
    vol.Exclusive(ATTR_ZONE_ARRAY_TEMPLATE, 'ZoneToClean'): cv.template,
    vol.Exclusive(ATTR_ZONE_REPEATER, 'TimeToRepeats'):
        vol.All(vol.Coerce(int), vol.Clamp(min=1, max=3)),
    vol.Exclusive(ATTR_ZONE_REPEATER_TEMPLATE, 'TimeToRepeats'): cv.template,
    vol.Optional('data'): dict,
    vol.Optional('data_template'): {cv.match_all: cv.template_complex},
})
```

Reply:
We should not include the optional template validators.

Reply:
> We should not include the optional template validators.


my test automation:
```
- id: '1546608481222'
  initial_state: 'off'
  alias: Test vacuum zone3
  trigger:
  - event: start
    platform: homeassistant
  condition: []
  action:
  - service: vacuum.xiaomi_clean_zone
    data_template:
      entity_id: vacuum.xiaomi_vacuum
      zone: [[8, 1, 1, 1],[1, 1, 1, 1]]
      repeats_template: '{{states.input_number.vacuum_passes.state|int}}'
```

my validation schema:
```
SERVICE_SCHEMA_CLEAN_ZONE = VACUUM_SERVICE_SCHEMA.extend({
    vol.Exclusive(ATTR_ZONE_ARRAY, 'ZoneToClean'):
        vol.All(list, [vol.ExactSequence([int, int, int, int])]),
    vol.Exclusive(ATTR_ZONE_ARRAY_TEMPLATE, 'ZoneToClean'): cv.template,
    vol.Exclusive(ATTR_ZONE_REPEATER, 'TimeToRepeats'):
        vol.All(vol.Coerce(int), vol.Clamp(min=1, max=3)),
    vol.Exclusive(ATTR_ZONE_REPEATER_TEMPLATE, 'TimeToRepeats'): cv.template,
})
```

Error received:
**Error while executing automation automation.test_vacuum_zone3. Invalid data for call_service at pos 1: expected int @ data['zone'][0]**

Reply:
Remove the `ATTR_ZONE_REPEATER_TEMPLATE` validation item.

Reply:
> Remove the `ATTR_ZONE_REPEATER_TEMPLATE` validation item.

Well! now is working!
Thank you so much!

Could you explain me how to pass list in template format?
If I use zone: '{{ [[5, 1, 1, 1],[1, 1, 1, 1]]  | list }}' i get error: Invalid data for call_service at pos 1: expected list for dictionary value @ data['zone']

Reply:
Something like this, I think:

```yaml
action:
  - service: vacuum.xiaomi_clean_zone
    data_template:
      entity_id: vacuum.xiaomi_vacuum_cleaner
      repeats: "{{ states('input_number.repeats') | int }}"
      zone:
        - - "{{ '5'|int }}"
          - "{{ '1'|int }}"
          - "{{ '1'|int }}"
          - "{{ '1'|int }}"
        - - "{{ '1'|int }}"
          - "{{ '1'|int }}"
          - "{{ '1'|int }}"
          - "{{ '1'|int }}"
```

Reply:
Fixed